### PR TITLE
Extract solana-poseidon crate from solana-program

### LINKED ISSUE #871: program: move poseidon module somewhere else
#### Problem
The light-poseidon dependency increases build time by a lot (25% last time I checked) and there is approximately one team that uses it.

### PR DESCRIPTION
**Pull request by kevinheavey**
#### Problem
See #871

#### Summary of Changes
This makes a new solana-poseidon crate and moves the poseidon module of solana-program there. I didn't re-export poseidon in Solana program because my understanding is that the number of teams using it is small.

This improves compile times by a lot: on my machine the time to build solana-program from scratch goes from 41 seconds to 32 seconds. What's more, the time to build solana-program after compiling all its dependencies goes from 14 seconds to 6 seconds.

Fixes #871

### CODE REVIEW

**Review by buffalojoec — COMMENTED** [review]
I like this change. Curious if you think `blake3` and `keccak` can also get similar treatment? If so, maybe we stick these under a folder `hashers` to minimize directory swelling?

File: poseidon/src/lib.rs
Comment (buffalojoec) [inline poseidon/src/lib.rs:1]: There's a call to `crate::syscalls::sol_poseidon` in this file that would need to get handled. Starting to think we might need a syscalls crate as well.

Comment (kevinheavey) [reply poseidon/src/lib.rs:1]: Fixed. And yeah at least one syscalls crate. When I played around with splitting up the SDK a while ago I ended up having a "core" syscalls crate with no dependencies: https://github.com/kevinheavey/agave/blob/smol-sdk/new_crates/syscall-core/Cargo.toml

**Comment by kevinheavey** [comment]
> I like this change. Curious if you think blake3 and keccak can also get similar treatment? If so, maybe we stick these under a folder hashers to minimize directory swelling?

@buffalojoec Yes those can be separate crates too. I don't really have an opinion on folder hierarchies except that I wish each workspace was in its own folder. Lmk if you want this PR to put the solana-poseidon crate in the hashers dir

**Review by buffalojoec — COMMENTED** [review]
Hey @kevinheavey sorry I took a bit to circle back to this. I wanted to ideate more on the `hashers` concept we discussed. However, even if we still decide to do that, we can always move this crate around.

I'm in favor of the change here. Let's see how CI does since it's been a while. Thanks for continuing to roll in dependency changes!

cc @samkim-crypto 

EDIT: I reserved the crate for now, and if there's consensus to add it I'll move under Anza. I don't know if CI will work until then, but at least sanity should.

**Comment by kevinheavey** [comment]
I think test-sanity.sh wants the solana-poseidon version to be 2.0.0, let's see how CI goes now

File: poseidon/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[package]
+name = "solana-poseidon"
+description = "Solana Poseidon hashing"
+documentation = "https://docs.rs/solana-poseidon"
+version = "2.0.0"
```
Comment (buffalojoec) [inline poseidon/Cargo.toml]:
```suggestion
version = { workspace = true }
```

File: poseidon/src/lib.rs [hunk 2]
```diff
@@ -258,7 +258,7 @@ pub fn hashv(
     {
         let mut hash_result = [0; HASH_BYTES];
         let result = unsafe {
-            crate::syscalls::sol_poseidon(
+            solana_program::syscalls::sol_poseidon(
```
Comment (buffalojoec) [inline poseidon/src/lib.rs]: This appears to be the only reason this crate needs to depend on `solana-program` at all.

I'm not 100% sure this is The Way™, but we could just implement this syscall binding manually in the crate.

```rust
extern "C" {
    pub fn sol_poseidon(
        parameters: u64,
        endianness: u64,
        vals: *const u8,
        val_len: u64,
        hash_result: *mut u8,
    ) -> u64;
}
```

Then this line just becomes

```suggestion
            sol_poseidon(
```

@CriesofCarrots would love your thoughts on this.

Comment (kevinheavey) [reply poseidon/src/lib.rs]: Sounds fine to me anyway!

**Review by buffalojoec — DISMISSED** [review]
This lgtm @kevinheavey, thanks! Let me get the crate set up and I'll loop back here.

**Review by buffalojoec — APPROVED** [review]

**Pull request merged**